### ISSUE TITLE
"Dialog" not announced by the screen reader when the modal opens

### ISSUE THREAD

**Issue by mlegait** [reactjs/react-modal#804]
### Steps to reproduce:

I created a simple CodePen using the code example in the README: https://codepen.io/mlegait/pen/eYpEWpz
I just replaced the `contentLabel` with `aria-labelledby`.

I'm on **Windows 10 Enterprise** with **NVDA 2019.3.1** and **Chrome 81.0.4044.129** (the last one).

### Expected behavior:

When testing the example provided by the W3C: https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html when the modal opens, it is announced by the screen reader like this: "[aria-labelledby text] dialog" so that the user knows that a dialog has been opened.

BUT in my CodePen example, it's just announced like this "[aria-labelledby text]" (without "dialog") so the user has no clue that a dialog has been opened.

### Link to example of issue:

https://codepen.io/mlegait/pen/eYpEWpz

### Additional notes:

I compared the attributes of the W3C dialog and the react-modal dialog and I don't see what the issue could be, maybe the element which is focused 🤔 

Thank you very much for your help 😄

**Comment by diasbruno** [comment]
Hi @mlegait, `role="dialog"` should be added by default in the "content" element of the modal.
Also, it can be changed using `role="alertdialog"`.

Can you please test with this aria attributes?

```js
aria={{
  labelledby: "id-modal-title",
  modal: "true"
}}
```

**Comment by mlegait** [comment]
Hi @diasbruno and thank you for your anwser :)

I tried adding `role="alertdialog"` to `ReactModal` and indeed it's announced by NVDA like this: "[aria-labelledby text] alert". It's good to know.
But in my use case, I need a dialog and not an alert.

When I inspect the code generated, the role dialog is well set:
```html
<div style="[...]" class="ReactModal__Content ReactModal__Content--after-open" tabindex="-1" role="dialog" aria-labelledby="id-modal-title" aria-modal="true">
```
But despite this, "dialog" is not announced by the screen reader :/

**Comment by diasbruno** [comment]
Hmm...if `role="dialog" aria-labelledby="id-modal-title" aria-modal="true"` is rendering correctly, I wonder if the `tabIndex` can cause problems in this case.

Perhaps, removing this line [ModalPortal.js#L362](https://github.com/reactjs/react-modal/blob/master/src/components/ModalPortal.js#L362) and see if it has something to do with this problem.

**Comment by mlegait** [comment]
Thanks.

How could I test if it fixes my issue? I guess I should install react-modal on my local env and run some included examples. 

Do you know why there is this `tabIndex="-1"`? Could we break something by removing it?

**Comment by diasbruno** [comment]
Unfortunately, I can't debug this (I don't have the necessary tools right now), but I can help.

> How could I test if it fixes my issue?

If you have a project using react-modal, you can patch directly on node_modules or using a fork.

> Do you know why there is this tabIndex="-1"?

`tabIndex` in this case is to avoid the element (`div` wrapping the content) to receive focus using "tab" - only via javascript `.focus()`.
This element receives focus automatically after the modal is open.

> Could we break something by removing it?

Hard to say, but, my guess is that it won't break anything. The modal controls the navigation (tab) inside of it and it uses the elements inside of that content element. So, this element must be excluded by default. And if it receives focus, than we have a bug on the navigation.

**Comment by mlegait** [comment]
I tried to remove `tabIndex: "-1"` in `ModalPortal.js` in the `node_modules` of one of my projects. And it's worse :/ Nothing is read anymore and the escape key doesn't work anymore to close the modal :(

I looked again at the example given by the W3C: https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html
There is indeed a difference with `tabindex`. There is no `tabindex` set on the dialog container but there is 2 divs with `tabindex="0"` around the dialog:
```html
<div tabindex="0"></div>
<div role="dialog" id="dialog1" aria-labelledby="dialog1_label" aria-modal="true" class="default_dialog">[...]</div>
<div tabindex="0"></div>
```
The other difference I noticed is that when the modal is open, the focus is set on the first element inside the modal (except for the title which is used by `aria-labelledby`) and not on the modal container as it is done by react-modal. 

Does it help?

**Comment by cp79shark** [comment]
The W3 sample has their dialog content as part of the initial DOM, `react-modal` is dynamically added content. `<h2>` cannot receive focus and yes per the guidelines the first focusable element in a dialog should be focused by default. Those two `div` elements are used to trap focus to the dialog by the dialog implementation on that page. https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/js/dialog.js

The issue seems to be because of the browser's accessibility tree not handling this scenario (labelledby with dynamic dom inserts). I'd say it's more an issue with NVDA than `react-modal`. You can see this play out in your pen modified to be the "rendered" dialog. https://codepen.io/cp79shark/pen/dyYdJNa

A workaround is to specify your label explicitly. `aria: { label: "Hello Dialog", modal: true }`. It's not ideal, but it's probably easier than getting NVDA to change things on their end and that still won't help users who haven't updated.

**Comment by mlegait** [comment]
I just installed **Jaws** and "Hello modal dialog" is well announced. So it seems you're right, it should be an NVDA bug.

Thank you so much for your help 😄

**Comment by techieshark** [comment]
> it seems you're right, it should be an NVDA bug.

Sounds like this can be closed yeah?

**Comment by mlegait** [comment]
Yes, it can be closed. 
Here is the NVDA related issue: https://github.com/nvaccess/nvda/issues/8620

Thank you 😄
